Highlight overdue/late payments in admin subscriptions list

haszari opened the pull request:
Fixes #260

## Description

Occasionally subscriptions can show as `Active` status when payment is overdue, in particular with offsite billing (e.g. WCPay Subscriptions using Stripe Billing). 

The goal of this PR is to highlight active subscriptions that may have overdue or missed payment issues. 

The existing `i` info button / tooltip component now shows red if payment is overdue. The tooltip message reads: `Subscription payment overdue`. 

This could help merchants take action, e.g. recover payment or suspend. 

Note this is a low-risk tweak to help mitigate this situation with minimal change to the existing UI. This issue is relatively rare in WCPay Subscriptions (estimated affecting 1-5% of merchants), so it's not worth making any more extensive changes right now.

Other options were considered, and could be improvements in future:

- Use a more modern icon (e.g. Gutenberg component).
- Add a new column for complex gateway-related info. Currently we're using a tooltip for two things (see below). This kind of info may be useful elsewhere in UI and/or for filtering.
   1. Indicate renewals are processed offsite – e.g. PayPal or WCPay Subscriptions.
   2. Indicate payment may be overdue (as of this PR). 
- Use a new, specific status when subscriptions are overdue for payment.
![Screenshot 2023-01-19 at 2 19 40 PM](https://user-images.githubusercontent.com/4167300/213333165-4c4141bb-6fc0-40b5-bf2c-bf7840049983.png)


Above screenshot shows both info (1 offsite renewal) and (2 overdue payment). Red icon is only used for overdue, and overdue payment message is displayed first and in bold, to disambiguate & highlight most important issue. When payment is overdue, the renewal mechanism (offsite vs tokenised onsite) is important info, so deliberately including that in tooltip when applicable.

## How to test this PR
1. Ensure you have some subscription products. Can use WC Subscriptions or WCPay Subscriptions. Tip: use daily subscription term so you have more renewal events to test with over time :) 
4. Purchase a subscription or set up a subscription. 
5. Prevent payment on renewal, or simulate this, for example:
   - Manual subscription - no auto renewal, will go overdue on renewal date.
   - Hack subscription next payment date in the past - set `_schedule_next_payment` post meta (aka subscription meta).
6. Go to `wp-admin > Dashboard > Subscriptions`. Any overdue subscriptions should show red icon.

## Product impact
<!-- What products will this PR ship in? -->

- [x] Added changelog entry (or does not apply)
- [x] Will this PR affect WooCommerce Subscriptions? yes/no/tbc, add issue ref
- [x] Will this PR affect WooCommerce Payments? yes/no/tbc, add issue ref
- n/a <!-- 🚨 Deprecations 🚨 --> Added deprecated functions, hooks or classes to the [spreadsheet](https://docs.google.com/spreadsheets/d/1xw9xszcPMnWsp4C8OKZMsLzZob7tOmWT7qMqmEIq314/edit#gid=0)

Commit:
show "overdue" in admin subscriptions list for failed renewals

Commit:
phpcs fixes - gmtime, strict equality

haszari commented:
Hitting pause on this PR. I'm switching approach, focus on adding logging and/or order notes for `suspend` / `pause collection` requests in WCPay client, to help diagnose the root cause affecting these stores.

haszari committed:
Merge branch 'trunk' into fix/260-highlight-failed-renewals

haszari commented:
>  Hitting pause on this PR. I'm switching approach, focus on adding logging and/or order notes for suspend / pause collection requests in WCPay client, to help diagnose the root cause affecting these stores.

Unblocked, will implement subtle red icon design: https://github.com/Automattic/woocommerce-subscriptions-core/issues/260#issuecomment-1387697410

Commit:
show red icon + tooltip for overdue subscriptions +
+ rename $date_value to hint type (unix timestamp) for readability

haszari commented:
Currently looks like this, using existing icon:

![Screenshot 2023-01-19 at 11 27 18 AM](https://user-images.githubusercontent.com/4167300/213308895-e6916410-fd8c-43c2-9481-a72cdada625f.png)


~~At a minimum I'm thinking we need to only show a single combined icon + tooltip in each row - tbc.~~ Updated 

Would be good to use a modern icon from wp components / gutenberg, will investigate options. However the existing icon + tooltip UX is most consistent so may move forward with that.

Commit:
render a single Next Renewal date tooltip:
- stack multiple messages
- use red icon for overdue payment

Commit:
embolden "overdue" tooltip message so it's clear

Commit:
changelog

haszari commented:
Requesting review from Helix and also @james-allan second opinion since he's up to speed on the wider context here.

haszari committed:
Merge branch 'trunk' into fix/260-highlight-failed-renewals

Commit:
fix equals-align phpcs issues

haszari commented on includes/admin/class-wcs-admin-post-types.php:
```diff
@@ -660,14 +660,33 @@ public static function get_date_column_content( $subscription, $column ) {
 			$column_content = '-';
 		} else {
-			$column_content = sprintf( '<time class="%s" title="%s">%s</time>', esc_attr( $column ), esc_attr( date( __( 'Y/m/d g:i:s A', 'woocommerce-subscriptions' ), $subscription->get_time( $date_type, 'site' ) ) ), esc_html( $subscription->get_date_to_display( $date_type ) ) );
+			$column_content = sprintf( '<time class="%s" title="%s">%s</time>', esc_attr( $column ), esc_attr( date( __( 'Y/m/d g:i:s A', 'woocommerce-subscriptions' ), $date_timestamp ) ), esc_html( $subscription->get_date_to_display( $date_type ) ) );
```
I'm getting a pesky phpcs error on this line. 

>  668 | ERROR   | [ ] date() is affected by runtime timezone changes which can cause date/time to be incorrectly displayed. Use gmdate() instead.
 
This is existing code – we're using `date()` already. This seems correct to me, we want to display the date in the server's timezone. I don't understand why this rule exists, and changing to `gmdate` would introduce a bug. 

Is there an alternative that allows us to explicitly request a localised date, e.g. something like `date( 'server_timezone' )`? Or do we want to turn this rule off?

james-allan replied:
Ah yes. 

The phpcs error is technically correct, we should never use date functions that are affected by runtime timezone changes. We've had pretty bad subscription issues in the distant past because folks change their PHP timezone for some reason and we were using functions like `date()`.

But the suggestion phpcs gives to use `gmdate()` raises some confusing date-related questions, that don't make a tonne of sense. 

---

The php manual says that `gmdate()` is: _"Identical to the [date()](https://www.php.net/manual/en/function.date.php) function except that the time returned is Greenwich Mean Time (GMT)."_ 

But IMO that is kinda misleading because in my understanding it doesn't manipulate the timestamp input at all to make it refer the GMT timezone equivalent. Which is why phpcs is recommending it. 

So, to be clear, calling `gmdate( '...', $timestamp )` where the timestamp param has already been adjusted to reflect the site's timezone, will display a date that reflects the site's timezone, it won't a GMT date at all. 

I hope that makes sense, but its super confusing.

james-allan replied:
Just to add more details.

We store subscription dates in the database in GMT (UTC+0). 
When you call `$subscription->get_time( ..., 'site' )` we take the date stored in GMT, convert it to a timestamp and then offset it based on the site's timezone offset. eg +10 hrs, -7 hrs etc. 

So for example: 

If I purchase a subscription at 1 pm brisbane time, the start date in the database is recorded as `2023-01-24 03:12:09` which is 13:12 AEST - 10 hrs = 03:12 UTC

If I then call `$subscription->get_time( ... )` I will get a timestamp that is 03:12. 

If I call `$subscription->get_time( ..., 'site' )` I will get a timestamp that is 13:12.

If we call `gmdate( ..., $subscription->get_time( ..., 'site' ) )` the date displayed will be 13:12 which is correct.

What we've done is passed it a site converted timestamp, it's basically the same as if we called `gmdate` with the timestamp equal to 13:12 in GMT.

I hope that makes sense, but to be really explicit, the phpcs is correct, we should just call `gmdate()` instead of `date()` here. It should be the same outcome.

haszari replied:
Great explanation! Makes sense, I was confused and thought that the localisation (timezone offset) provided by `gmdate` was valid, I see now that it doesn't rely on consistent settings. I guess there's a possibility this fixes a bug, some sites might have seen different dates.

haszari commented on includes/admin/class-wcs-admin-post-types.php:
```diff
@@ -660,14 +660,33 @@ public static function get_date_column_content( $subscription, $column ) {
+				$tooltip_message = '';
+				$tooltip_classes = 'woocommerce-help-tip';
+
+				if ( $subscription_is_active && $date_timestamp < current_time( 'timestamp', false ) ) {
```
This is a phpcs warning, though it sounds possibly relevant, could be causing a bug.

>  677 | WARNING | [ ] Calling current_time() with a $type of "timestamp" or "U" is strongly discouraged as it will not return a Unix (UTC) timestamp. Please consider using a non-timestamp format or otherwise refactoring this code.
 
What I want to do here is compare the subscription last payment date (timestamp) with the current time (in timestamp format). In my testing this is working correctly (though maybe I missed a test case). 

What's the best way to compare dates (in WordPress/PHP) – what do we do elsewhere in the code?

And does that meet the phpcs standards?

I see that phpcs is stricter than our current code base, so whenever we touch code we'll need to address phpcs issues. Need to take care not to introduce bugs :)

james-allan replied:
> What's the best way to compare dates (in WordPress/PHP) – what do we do elsewhere in the code?

The issue in this case is that the `$date_timestamp` is a timestamp in the site's timezone. 

I don't think we have any instance where we're trying to determine if a site-converted-timestamp is in the past. We are usually comparing a UTC/GMT one in which case we use either: 

- `time()`
- `gmdate( 'U' )` - GMT timestamp for the current time. 

---

I think there are 3 ways to resolve this:

1. Keep the timestamp variable `$date_timestamp` as a site timestamp. Then to determine if it's in the past you'll have to do some manipulation like this: `$date_timestamp < ( time() + wc_timezone_offset() )`. Which is basically offsetting the timestamp returned by `time()` so we're comparing like for like - two site timestamps.
2. Rather than trying to do both date-displaying (which is typically a date in the site's timezone) and date-logic (which is typically done in UTC) with the same date variable, you could split it. eg: 
   - use `$subscription->get_time( ... ) < time()` for logic conditions and 
   - use `$subscription->get_time( ..., 'site' )` above where it's displayed to the user.
3. Use the `WC_DateTime` object which has functions for easy access to both a site and UTC timestamps. I've included a couple of code snippets example below to illustrate how you could do it: 

Replace your `$date_timestamp = $subscription->get_time( $date_type, 'site' );` with this
```php
// Get a WC_Datetime instance - the timestamp passed in must be UTC. 
$datetime = wcs_get_datetime_from( $subscription->get_time( $date_type ) );
```

Replace the `date( __( 'Y/m/d g:i:s A', 'woocommerce-subscriptions' ), $date_timestamp )` above with the following

```php
// This will display the date in the site's timezone given the specified format.
$datetime->date_i18n( __( 'Y/m/d g:i:s A', 'woocommerce-subscriptions' ) );
```

Finally, replace this `$date_timestamp < current_time( 'timestamp', false )` condition with:

```php
// getTimestamp returns the UTC timestamp.
$datetime->getTimestamp() < time();
```

haszari replied:
Nice! TIL `DateTime` class, this is much clearer and more explicit.

```php
$column_content = sprintf( 
  '<time class="%s" title="%s">%s</time>', 
  esc_attr( $column ), 
  esc_attr( $datetime->date_i18n( __( 'Y/m/d g:i:s A', 'woocommerce-subscriptions' ) ) ), 
  esc_html( $subscription->get_date_to_display( $date_type ) ) 
);
```

Side note, I see we are rendering the date twice on the column - once for content, once for `title` attr. I'm tempted to split these out to variables to document why each format is slightly different. Do you recall why they are different?

james-allan replied:
The title format is a raw date and the `get_date_to_display()` can return a relative date.

So the idea is you can hover the relative date and see the actual raw date. 

<img width="225" alt="Screen Shot 2023-01-25 at 9 33 33 am" src="https://user-images.githubusercontent.com/8490476/214444544-5cb4d300-5f8b-4880-8015-cf388b0bc1ec.png">

haszari replied:
Aha - similar to GitHub comment timestamps! I did a little refactor to capture that intention: b279a31a7519193fd1bbac32735124cc3953dcde

Commit:
Merge remote-tracking branch 'origin/fix/260-highlight-failed-renewals' into fix/260-highlight-failed-renewals

Commit:
use DateTime class to tidy date comparisons and address phpcs issues:
- use DateTime instance for local variable
- use DateTime::date_i18n to render to localised string for display
- use DateTime::getTimestamp to compare with now / determine if in past

Commit:
refactor: clarify different date formats used in list view

haszari commented:
Thanks for the instructive feedback @james-allan  - ready for another review/test, if all good feel free to merge if I'm not around.

james-allan commented:
Changes look good to me and test well. In line with your comment above, I'll go ahead and merge this. 😄

Jinksi commented on changelog.txt:
```diff
@@ -2,6 +2,7 @@
 
 = 5.2.0 - 2022-xx-xx =
 * Add - New wcs_set_order_address() helper function to set an array of address fields on an order or subscription.
+* Add - Highlight subscriptions with overdue payment in list view with red icon & tooltip.
```
This changelog entry should probably be moved to `5.3.0` since [`5.2.0` has already been tagged](https://github.com/Automattic/woocommerce-subscriptions-core/releases/tag/5.2.0).

haszari replied:
Good spotting - thanks @Jinksi ! Would this get caught & fixed during subs-core tag/"release" process?

james-allan replied:
> Would this get caught & fixed during subs-core tag/"release" process?

Nah it wouldn't automatically be apparent that this changelog entry was added after 5.2.0 was tagged. :( 

I can just submit a small PR to move it up.

haszari replied:
Thanks James!

> Nah it wouldn't automatically be apparent that this changelog entry was added after 5.2.0 was tagged. :(

Dang, that's a tiny bit fragile. I think I personally look at a diff of the changelog when I did (a while ago?!) releases, which might catch it. IMO a good release process has one person eyeball the full changelog before it goes out, to catch duplicates, missing stuff, inconsistent wording etc.

haszari replied:
A diff like this is a good way to eyeball changelog changes before tagging: https://github.com/Automattic/woocommerce-subscriptions-core/compare/5.2.0...trunk#diff-b40cd67182487ef24807d9c9268329d35fbd96aa2b0a9cae69e2e0d746b1c666

james-allan replied:
> Dang, that's a tiny bit fragile.

Yeah. 

Doing a mini-retro personally, I was thinking about how this happened. I put it down to the changelog in this PR's changeset shows the `5.2.0 22-xx-xx` which indicates it hasn't been tagged. The fact that it had indicates this branch would have been behind `trunk` and so I must have missed the "needs update" flag on this PR as well. 

So there are a few things that occurred that would have caught this. Updating the branch before merging and paying more attention to the version. 

> I think I personally look at a diff of the changelog when I did (a while ago?!) releases, which might catch it. IMO a good release process has one person eyeball the full changelog before it goes out, to catch duplicates, missing stuff, inconsistent wording etc.

This probably isn't the right place for this discussion to occur but I'd prefer we moved towards using Jetpack changelogs over introducing a manual process for releases which has fragility issues of its own.
